### PR TITLE
[Flang][OpenMP] Prepare lowering for composite construct support

### PR DESCRIPTION
**Pull request by skatrak**
This patch performs various changes in preparation for adding lowering support for composite constructs. In particular, the following changes are done:
- Splitting lowering for `omp.loop_nest` into its own function and update lowering for all supported loop wrappers to not create this operation themselves.
- Creating "wrapper" lowering functions called by their associated "standalone" lowering functions, which also are responsible for calling the new codegen function for `omp.loop_nest`. "Wrapper" lowering functions will later be reused by "composite" lowering functions as well.
- Implementing the creation of a single `DataSharingProcessor` inside of `genOMPDispatch` to be used while lowering standalone and composite loop-associated constructs. This avoids code duplication and ensures that no privatization-related allocations are created inside of loop wrapper regions.
- Preventing the `FirOpBuilder::getAllocaBlock` method from returning blocks of a loop wrapper, as inserting allocations there would violate existing requirements on the contents of a loop wrapper region.
- Removing the introduction of `fir.undef` operations as a way to keep track of insertion points inside of the `DataSharingProcessor` to avoid leaving such operations inside of loop wrappers.
- Updating tests broken by reordering of privatization with respect to clause processing.

### CODE REVIEW

File: flang/lib/Optimizer/Builder/FIRBuilder.cpp
Comment (kparzysz) [inline flang/lib/Optimizer/Builder/FIRBuilder.cpp:1]: This could be its own PR.

**Review by kparzysz — COMMENTED** [review]
The changes look ok to me, but there are several things bundled together.  It would make it a lot easier to review if the parts that could be done in isolation were in separated PRs (perhaps forming stacks).

**Comment by skatrak** [comment]
Thank you @kparzysz for the comment. I'll close this and open a few PRs in a stack to make reviews a bit easier.
